Updated checkout

alfredog1976 opened the pull request:
- New reliable click buy button code
- Pass in cvv, shipping address, shipping option
- shoe-size no long required (due to special shoe type, such as Air Presto)
- shoe-type handles special shoe type. 
- Exception thrown if "W", "Y", "C" passed in without shoe size.
- Poll for which checkout section appears first
- Poll after shipping option selection for which checkout section appears next (for user who hasn't previously purchased a SNKRs release vs a user who has)
- Added wait_until_present
- xpath used for save button (payment section) and submit button (purchase) due to potential iFrame issue if cvv has been input. 
- Added StaleElementReferenceException for 2nd save button click (shipping option) to attempt retry save button click once

alfredog1976 committed:
Updates to match Nike site

Login fixed as well as shoe size and type selection. Also added new parameters: shoe-type and dont-quit. Also, refresh page if shoe size and type selection fails.

alfredog1976 committed:
Updated main.py with suggestions for shoe_type

Removed shoe_type command line variations. Replaced original if Youth/Child check with "if shoe_type in ("Y", "C"):"

alfredog1976 committed:
Determine size only vs size + size type

Determine size only vs size + size type by checking text of size available first element. Uses regular expression to determine which code logic to use. Size only uses part of code originally used to determine size.

alfredog1976 committed:
Update main.py

alfredog1976 committed:
Merge branch 'master' into master

alfredog1976 committed:
scrollIntoView Buy button

scrollIntoView exception occurs when browser window not big enough and Buy button is not in view. Attempts to scroll next element (class="photo-component") since there were issues trying to scroll Buy button element into view. This has only been tested with Firefox.

alfredog1976 committed:
Revert

revert

alfredog1976 committed:
Revert "Revert"

This reverts commit d3fb3880339a04a86ea8d776226eae77a95a3705.

alfredog1976 committed:
Merge branch 'master' of https://github.com/alexschimpf/Snkrs-Bot into alexschimpf-master
merge from master

alfredog1976 committed:
Merge pull request #2 from alfredog1976/alexschimpf-master

Merge from master

alfredog1976 committed:
Update main.py

alfredog1976 committed:
Update README.md

alfredog1976 committed:
Update main.py

alfredog1976 committed:
Removed random sleep

Removed random sleep from "retry_login"

alfredog1976 committed:
Updated checkout

- New reliable click buy button code
- Pass in cvv, shipping address, shipping option
- shoe-size no long required (due to special shoe type, such as Air Presto)
- shoe-type handles special shoe type. Exception thrown if "W", "Y", "C" passed in without shoe size.
- Poll for which checkout section appears first
- Poll after shipping option selection for which checkout section appears next
- Added wait_until_present
- xpath used for save button (payment section) and submit button (purchase) due to potential iFrame issue if cvv inputted.
- Added StaleElementReferenceException for 2nd save button click (shipping option) to retry save button click once

alfredog1976 commented:
Updated code thoroughly tested and was able to reach “Purchased shoe” in both headless and browser mode
![AA656E92-22F4-4223-ABFF-712DA97AFB70](https://user-images.githubusercontent.com/28766561/100893385-2b28ae80-3470-11eb-8938-937155f14458.png)

alexschimpf commented:
@alfredog1976 Damn, nice work. Gonna take me some time to review it. In the meantime, you've got a merge conflict to fix.

alfredog1976 committed:
Merge branch 'master' into master

alexschimpf commented on README.md:
```diff
@@ -78,12 +78,27 @@ Here is a list and description of the different arguments to use for the script:
 
 <b>--url</b>
 * URL for desired shoe
+* Size parameter can also be passed in (for example: https://www.nike.com/launch/t/kobe-5-protro-bruce-lee?size=11). In this case, `--shoe-size` and `--shoe-type` will be ignored
```
So is the advantage of using the `size` query param that it's less likely to fail since that size will already be pre-selected? Also wondering if `--shoe-type` can be expressed as a query param. I don't love the idea of allowing for both ways (especially when a combination of using both can cause issues). You could make it so that the `--shoe-size` param is appended to the `--url` instead of doing xpath to find the size button.

alfredog1976 replied:
Can only pass size as a query parameter. Wanted to give user flexibility to choose either option. My assumption was the query parameter would perform slightly faster than passing command line parameter and searching the DOM size selection section. (Not sure what Nike is doing exactly behind the scenes and I didn’t have time to run a performance analysis comparing each approach). As I noted in the ReadMe, passing in just the size as a query parameter on a release that has multiple shoes on he same page will result in an unintended shoe size & type selection.

Another observation is when users copy the url, the size query parameter is also there (I believe Nike appends it if the size is set in the user’s Nike profile).

So, options are: leave as is or use your approach, but we’ll need to do an additional check to see if the size-type command line parameter has been passed in.

alexschimpf replied:
Alright, let's just leave it as is.

alexschimpf commented on main.py:
```diff
@@ -345,8 +652,12 @@ def wait_until_visible(driver, xpath=None, class_name=None, duration=10000, freq
-    run(driver=driver, shoe_type=shoe_type, username=args.username, password=args.password, url=args.url, shoe_size=args.shoe_size,
-        login_time=args.login_time, release_time=args.release_time, page_load_timeout=args.page_load_timeout,
+    run(driver=driver, shoe_type=shoe_type, username=args.username, password=args.password, url=args.url, shoe_size=args.shoe_size, shipping_option=args.shipping_option,
+        login_time=args.login_time, release_time=args.release_time, shipping_address=args.shipping_address, page_load_timeout=args.page_load_timeout,
```
You should `json.loads(shipping_address)` before passing it to `run`.

alexschimpf commented on main.py:
```diff
@@ -83,34 +92,171 @@ def run(driver, shoe_type, username, password, url, shoe_size, login_time=None,
+                    LOGGER.info("Too many iterations through checkout element check. Terminating check...")
+                    break
+                    
+            if skip_add_address is False and shipping_address:
```
This function is getting unwieldy and hard to read. Can you break some of this logic up into smaller functions?

alexschimpf commented on main.py:
```diff
@@ -83,34 +92,171 @@ def run(driver, shoe_type, username, password, url, shoe_size, login_time=None,
+                    LOGGER.exception("Failed to click save button: " + str(e))
+                    six.reraise(Exception, e, sys.exc_info()[2])
+                    
+                if retry_save:
```
I think you can just get rid of `retry_save` entirely and do this logic where you do `retry_save = True`.

alfredog1976 committed:
Updated checkout (review)

-Removed retry_save and placed code within exception handler
-Placed polling checkout section code in 2 functions (poll_checkout_phase_one and poll_checkout_phase_two)
- json.loads called before "run"

alexschimpf commented on main.py:
```diff
@@ -61,18 +66,21 @@ def run(driver, shoe_type, username, password, url, shoe_size, login_time=None,
 
+    skip_add_address = False
+    skip_select_shipping = False
+    skip_payment = False
```
Do we need these vars declared outside the loop? Like if you set them in one iteration, don't they need to be unset before the next iteration (if the first iteration failed)?

alfredog1976 replied:
You are correct. I have reset the variables after num_retries_attempted gets incremented.

alexschimpf commented on main.py:
```diff
@@ -83,34 +91,93 @@ def run(driver, shoe_type, username, password, url, shoe_size, login_time=None,
-
-            if select_payment:
+                
+            skip_add_address, skip_select_shipping, skip_payment = poll_checkout_phase_one(driver=driver, skip_add_address=skip_add_address, skip_select_shipping=skip_select_shipping, skip_payment=skip_payment)
```
`skip_add_address`, `skip_select_shipping`, and `skip_payment` should not be params of `poll_checkout_phase_one`. They are always False when passed in so they can just be local variables inside that function.

alfredog1976 committed:
Update skip vars

- Removed skip booplean variables as parameters to poll_checkout_phase_one and poll_checkout)_phase_two functions
- Added those variables as local variables within poll_checkout_phase_one and poll_checkout)_phase_two functions
- Reset skip boolean variables to 'False' in Exception section of code before next iteration of loop

symourjr commented:
I tried this on an existing launch product and it can't seem to find the buy button. On the page itself it says it's add to cart  and not buy, is this because it's not a drop?

symourjr commented:
Tried this out today. Got stuck on cvv input. Was able to save the html page though.

mnai01 commented:
> Tried this out today. Got stuck on cvv input. Was able to save the html page though.

Can you upload the HTML page

symourjr commented:
Once I find out how to sanitize some info from it.

symourjr commented:
Figured it out. Missing "cvv=args.cvv" on the run function.

alfredog1976 committed:
Added cvv

Added cvv parameter. Was in original code, mistakenly not included during merge

JThistle9 commented:
What's up boys, was trying to join in on the work but couldn't get past login page. We got a discord to discuss setup?

jjarrett21 commented:
Wanting to help on this project if yall are still working on it?
